compiler/rustc_errors/src/lib.rs

```diff
@@ -1000,6 +1003,8 @@ impl HandlerInner {
         } else {
             self.bump_warn_count();
         }
+
+        ErrorReported::new_i_know_what_im_doing()
```

Comment:
This needs to check `diagnostic.is_error()`.

Reply:
Overall, it feels like the nice thing would be to seal `ErrorReported` and have an `impl` in the same "sealed" module that `ErrorReported` would be in, with a wrapper method for `self.emitter.emit_diagnostic(diagnostic);` that checks for the conditions for `ErrorReported` and creates it if necessary.

Another wrapper could be added for `self.delayed_span_bugs.push(diagnostic);`. Those should be the only cases in which you can guarantee `ErrorReported`. Error deduplication can keep a `ErrorReported` around for reuse (in general, this should be favored because it's free, being a ZST).